src/analysis/redundant_attributes.d

```diff
@@ -0,0 +1,267 @@
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+
+module analysis.redundant_attributes;
+
+import dparse.ast;
+import dparse.lexer;
+import dsymbol.scope_ : Scope;
+import analysis.base;
+import analysis.helpers;
+
+import std.algorithm;
+import std.conv : to, text;
+import std.range : empty, front, walkLength;
+
+/**
+ * Checks for redundant attributes. At the moment only access specifiers.
+ */
+class RedundantAttributesCheck : BaseAnalyzer
+{
+	this(string fileName, const(Scope)* sc, bool skipTests = false)
+	{
+		super(fileName, sc, skipTests);
+		stack.length = 0;
+	}
+
+	override void visit(const Declaration decl)
+	{
+
+		// labels, e.g. private:
+		if (auto attr = decl.attributeDeclaration)
+		{
+			if (filterAttributes(attr.attribute))
+			{
+				addAttribute(attr.attribute);
+			}
+		}
+
+		auto attributes = decl.attributes.filter!(a => filterAttributes(a));
+		if (attributes.walkLength > 0) {
+
+			// new scope: private { }
+			if (decl.declarations.length > 0) {
+				auto prev = currentAttributes[];
+				// append to current scope and reset once block is left
+				foreach (attr; attributes)
+					addAttribute(attr);
+
+				scope(exit) currentAttributes = prev;
+				decl.accept(this);
+			} // declarations, e.g. private int ...
+			else
+			{
+				foreach (attr; attributes)
+					checkAttribute(attr);
+
+				decl.accept(this);
+			}
+		} else {
+			decl.accept(this);
+		}
+	}
+
+	alias visit = BaseAnalyzer.visit;
+
+	mixin ScopedVisit!Module;
+	mixin ScopedVisit!BlockStatement;
+	mixin ScopedVisit!StructBody;
+	mixin ScopedVisit!CaseStatement;
+	mixin ScopedVisit!ForStatement;
+	mixin ScopedVisit!IfStatement;
+	mixin ScopedVisit!TemplateDeclaration;
+	mixin ScopedVisit!ConditionalDeclaration;
+
+private:
+
+	alias ConstAttribute = const Attribute;
+	alias CurrentScope = ConstAttribute[];
+	ref CurrentScope currentAttributes() {
+		return stack[$ - 1];
+	}
+	CurrentScope[] stack;
+
+	void addAttribute(const Attribute attr)
+	{
+		removeOverwrite(attr);
+		if (checkAttribute(attr)) {
+			currentAttributes ~= attr;
+		}
+	}
+
+	bool checkAttribute(const Attribute attr)
+	{
+		auto match = currentAttributes.find!(a => a.attribute.type == attr.attribute.type);
+		if (!match.empty)
+		{
+			auto token = attr.attribute;
+			addErrorMessage(token.line, token.column, KEY,
+					text("same access specifier used as defined on line ",
```

Comment:
"access specifier" is inaccurate, it's called a [visibility attribute](https://dlang.org/spec/attribute.html#VisibilityAttribute).
If you add support for other attribute you'll have to think to use the correct term too.